pages/advanced/eip-7702/7702-safe.mdx

```diff
@@ -0,0 +1,59 @@
+import { Callout } from 'nextra/components'
+
+# Safe and EIP-7702
+
+EIP-7702 does not specify how to initialise the storage of the account but only gives a way to set the code of the account. This means that the account will be created with an empty storage, and the user will have to set the storage manually.
+
+Existing Safe contracts cannot be used with EIP-7702, because of following reasons:
+
+- Delegating to Safe Singleton or the Proxy contract will expose the EOA account to the risk of front-running during setup.
+- In its current implementation, the Safe Singleton contract does not let itself to become an owner meaning that after delegating to the Safe Singleton, the EOA account cannot sign Safe transactions and will need to keep another private key to do so.
+
+## Possible approaches
+
+### Modified safe proxy
+
+[SafeEIP7702Proxy](https://github.com/5afe/safe-eip7702/blob/main/safe-eip7702-contracts/contracts/SafeEIP7702Proxy.sol)
+
+This approach uses a derived proxy contract from [Safe Proxy](https://github.com/5afe/safe-eip7702/blob/main/safe-eip7702-contracts/contracts/SafeEIP7702Proxy.sol) with following changes:
+1. The constructor of the `SafeEIP7702Proxy` contract has an additional parameter `setupDataHash` which is the hash of the `setup` function call data. Thus, the address of the proxy contract also depends on the `setupDataHash` and not just the `singleton` contract address.
+Proxy contract uses this hash to verify that the `setup` function parameter values are unchanged during initialised of storage.
+
+2. The proxy implements `setup` function which calls the `setup` function of the `singleton` contract and has additional logic:
+- Set the storage slot 0 that is, the address of the singleton in the EOA storage.
+- Verify that the `setupDataHash` is equal to the hash of the `setup` function call data.
```

Comment:
- The constructor of the `SafeEIP7702Proxy` contract has the additional `setupDataHash` parameter, which is the hash of the `setup` function call data. Thus, the address of the proxy contract also depends on the `setupDataHash` and not just the Singleton contract address. The proxy contract uses this hash to verify that the `setup` function parameter values are unchanged during the initialization of storage.
- The proxy implements the `setup` function, which calls the `setup` function of the `singleton` contract and has additional logic:
  - Set the storage slot 0, that is, the address of the singleton in the EOA storage.
  - Verify that the `setupDataHash` equals the hash of the `setup` function call data.